### PR TITLE
CONTRIBUTING: Add a more fleshed out contributing guide

### PR DESCRIPTION
**Pull request by jschipp-r7**
## Proposed Changes

Describe the proposed changes:

  - Add a contributing guide based on Metasploit's excellent guide

## Testing

```
$ mdl CONTRIBUTING.md | grep -v 'MD013 Line length' && printf "We're good"

A detailed description of the rules is available at https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md
We're good
```

## Display

# Contributing

Thank you for your interest in improving InsightConnect! Please review our [Code of Conduct] before making contributions.

There are multiple ways to contribute beyond writing code. These include:

- [Submit bugs and feature requests] with detailed information about your issue or idea.
- [Help fellow users with open issues] or [help fellow committers test recent pull requests].
- [Report a security vulnerability in InsightConnect's plugins] to Rapid7.
- Submit an updated or brand new plugin!  We are always eager for new
  integrations or features. Don't know where to start? Check out the [developer documentation].

Here is a short list of do's and don'ts to make sure *your* valuable contributions actually make
it into production.  If you do not care to follow these rules, your contribution
**will** be closed. Sorry!

## Code Contributions

- **Do** read the [developer documentation]
- **Do** stick to the [Python PEP8] and [plugin style] guides.
- **Do** follow the [50/72 rule] for Git commit messages.
- **Do** license your code as MIT.
- **Do** create a [topic branch] to work on. This helps ensure users are aware of commits on the branch being considered for merge, allows for a location for more commits to be offered without mingling with other contributor changes, and allows contributors to make progress while a PR is still being reviewed.

### Pull Requests

- **Do** write "WIP" on your PR and/or open a [draft PR] if submitting **working** yet unfinished code.
- **Do** target your pull request to the **master branch**.
- **Do** specify a descriptive title to make searching for your pull request easier e.g. "Okta: add Suspend User action".
- **Do** include [console output], especially the JSON output for new features and bug fixes.
- **Do** list [verification steps] so your tests are reproducible.
- **Do** [reference associated issues] in your pull request description.
- **Don't** leave your pull request description blank.
- **Don't** abandon your pull request. Being responsive helps us land your code faster.

#### New Features

- **Do** install validator dependencies necessary to run `make validate` to find and fix any errors or warnings that come up.
- **Don't** include more than one plugin per pull request.
- **Do** include documentation showing sample run-throughs.

#### Bug Fixes

- **Do** include reproduction steps in the form of [verification steps].
- **Do** link to any corresponding [Issues] in the format of `See #1234` in your commit description.

## Bug Reports

Please report vulnerabilities in Rapid7 software directly to security@rapid7.com.
For more on our disclosure policy and Rapid7's approach to coordinated disclosure, [head over here](https://www.rapid7.com/security).

When reporting issues:

- **Do** write a detailed description of your bug and use a descriptive title.
- **Do** include reproduction steps, stack traces, and anything that might help us fix your bug.
- **Don't** file duplicate reports; search for your bug before filing a new report.

If you need additional guidance, reach out to the open source contribution owners at
`IntegrationAlliance@rapid7.com`

Finally, **thank you** for taking the few moments to read this far! You're already way ahead of the
curve, so keep it up!

[Code of Conduct]:./CODE_OF_CONDUCT.md
[developer documentation]:https://komand.github.io/python/start.html
[Submit bugs and feature requests]:https://github.com/rapid7/insightconnect-plugins/issues
[Report a security vulnerability in InsightConnect itself or its plugins]:https://www.rapid7.com/disclosure.jsp
[Help fellow users with open issues]:https://github.com/rapid7/insightconnect-plugins/issues
[help fellow committers test recent pull requests]:https://github.com/rapid7/insightconnect-plugins/pulls
[Python PEP8]:https://www.python.org/dev/peps/pep-0008/
[plugin style]:https://komand.github.io/python/style.html
[50/72 rule]:http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[Report a security vulnerability in Metasploit itself]:https://www.rapid7.com/disclosure.jsp
[topic branch]:http://git-scm.com/book/en/Git-Branching-Branching-Workflows#Topic-Branches
[draft PR]:https://help.github.com/en/articles/about-pull-requests#draft-pull-requests
[console output]:https://help.github.com/articles/github-flavored-markdown#fenced-code-blocks
[verification steps]:https://help.github.com/articles/writing-on-github#task-lists
[reference associated issues]:https://github.com/blog/1506-closing-issues-via-pull-requests
[Issues]:https://github.com/rapid7/insightconnect-plugins/issues

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Contributing
+
+Thanks for your interest in making InsightConnect a better security orchestration tool.
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]:
```suggestion
Thank you for your interest in making InsightConnect a better security orchestration tool. Please review our [Code of Conduct] before making contributions.
```
More formal

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Before you get started, review our [Code of Conduct].
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]:
```suggestion
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+There are multiple ways to help beyond just writing code:
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]:
```suggestion
There are multiple ways to contribute beyond writing code. These include:
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+- [Submit bugs and feature requests] with detailed information about your issue or idea.
+- [Help fellow users with open issues] or [help fellow committers test recent pull requests].
+- [Report a security vulnerability in InsightConnect itself or its plugins] to Rapid7.
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]: The scope of this repository is only plugins - I'm not sure I would mention the InsightConnect product

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+- Submit an updated or brand new plugin!  We are always eager for new
+  integrations or features. Don't know where to start? Check out the [developer documentation].
+
+Here's a short list of do's and don'ts to make sure *your* valuable contributions actually make
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]:
```suggestion
Here is a short list of do's and don'ts to make sure *your* valuable contributions actually make
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+it into the master branch.  If you do not care to follow these rules, your contribution
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]:
```suggestion
it into production.  If you do not care to follow these rules, your contribution
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+**will** be closed. Sorry!
+
+## Code Contributions
+
+- **Do** read the [developer documentation]
+- **Do** stick to the [Python PEP8] and [plugin style] guides.
+- **Do** follow the [50/72 rule] for Git commit messages.
+- **Do** license your code as MIT.
+- **Do** create a [topic branch] to work on instead of working directly on `master`. This helps protect the process, ensures users are aware of commits on the branch being considered for merge, allows for a location for more commits to be offered without mingling with other contributor changes, and allows contributors to make progress while a PR is still being reviewed.
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]: I'm not sure the first part of this statement is necessary. We *should* have merge protection enabled on `master` making it impossible to merge into it without going through a PR process which requires a branch.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+### Pull Requests
+
+- **Do** write "WIP" on your PR and/or open a [draft PR] if submitting **working** yet unfinished code.
+- **Do** target your pull request to the **master branch**.
+- **Do** specify a descriptive title to make searching for your pull request easier.
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]: How about requiring a format for this? Ie. "Plugin Name - Brief description of change"

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+- **Do** include [console output], especially the JSON output for new features and bug fixes.
+- **Do** list [verification steps] so your code is testable.
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]:
```suggestion
- **Do** list [verification steps] so your tests are reproducible.
```
?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+- **Do** [reference associated issues] in your pull request description.
+- **Don't** leave your pull request description blank.
+- **Don't** abandon your pull request. Being responsive helps us land your code faster.
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]: Would there be any benefit to cleaning out inactive pull requests after x days?

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+Please report vulnerabilities in Rapid7 software directly to security@rapid7.com.
+For more on our disclosure policy and Rapid7's approach to coordinated disclosure, [head over here](https://www.rapid7.com/security).
+
+When reporting issues:
+
+- **Do** write a detailed description of your bug and use a descriptive title.
+- **Do** include reproduction steps, stack traces, and anything that might help us fix your bug.
+- **Don't** file duplicate reports; search for your bug before filing a new report.
+
+If you need some more guidance, talk to the main body of open source contributors over at
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]:
```suggestion
If you need additional guidance, reach out to the open source contribution owners at
```

File: CONTRIBUTING.md (hunk 8 above)
Comment (jschipp-r7) [reply CONTRIBUTING.md]: Good idea

File: CONTRIBUTING.md (hunk 10 above)
Comment (jschipp-r7) [reply CONTRIBUTING.md]: We can think about this

File: CONTRIBUTING.md (hunk 8 above)
Comment (jschipp-r7) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md (hunk 7 above)
Comment (jschipp-r7) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md (hunk 4 above)
Comment (jschipp-r7) [reply CONTRIBUTING.md]: Done

File: CODE_OF_CONDUCT.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
+# Contributor Code of Conduct
+
+As contributors and maintainers of this project, and in the interest of
+fostering an open and welcoming community, we pledge to respect all people who
+contribute through reporting issues, posting feature requests, updating
+documentation, submitting pull requests or patches, and other activities.
```
Comment (mrinehart-r7) [inline CODE_OF_CONDUCT.md]:
```suggestion
documentation, submitting pull requests, and other activities.
```

File: CODE_OF_CONDUCT.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+We are committed to making participation in this project a harassment-free
```
Comment (mrinehart-r7) [inline CODE_OF_CONDUCT.md]: I'm not sure if lines 8 - 21 came from legal, but if not - run this by them. We don't want to open ourselves up for any damage with potential botched wording.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+There are multiple ways to contribute beyond writing code. These include:
+
+- [Submit bugs and feature requests] with detailed information about your issue or idea.
+- [Help fellow users with open issues] or [help fellow committers test recent pull requests].
+- [Report a security vulnerability in InsightConnect's plugins] to Rapid7.
+- Submit an updated or brand new plugin!  We are always eager for new
+  integrations or features. Don't know where to start? Check out the [developer documentation].
+
+Here is a short list of do's and don'ts to make sure *your* valuable contributions actually make
+it into production.  If you do not care to follow these rules, your contribution
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]:
```suggestion
it into production.  If you do not care to follow these rules, your contribution will be rejected
```
Using `closed` implies a pull request is used. Might want to play around with the wording.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+- **Do** read the [developer documentation]
+- **Do** stick to the [Python PEP8] and [plugin style] guides.
+- **Do** follow the [50/72 rule] for Git commit messages.
+- **Do** license your code as MIT.
+- **Do** create a [topic branch] to work on. This helps ensure users are aware of commits on the branch being considered for merge, allows for a location for more commits to be offered without mingling with other contributor changes, and allows contributors to make progress while a PR is still being reviewed.
+
+### Pull Requests
+
+- **Do** write "WIP" on your PR and/or open a [draft PR] if submitting **working** yet unfinished code.
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]: Why does the code have to be working? I would omit that - WIP is totally valid for non-functional, but unfinished code.

Comment (mrinehart-r7) [reply CONTRIBUTING.md]: Example: someone just wants eyes on it or is requesting assistance.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+- **Do** list [verification steps] so your tests are reproducible.
+- **Do** [reference associated issues] in your pull request description.
+- **Don't** leave your pull request description blank.
+- **Don't** abandon your pull request. Being responsive helps us land your code faster.
+
+#### New Features
+
+- **Do** install validator dependencies necessary to run `make validate` to find and fix any errors or warnings that come up.
+- **Don't** include more than one plugin per pull request.
+- **Do** include documentation showing sample run-throughs.
```
Comment (mrinehart-r7) [inline CONTRIBUTING.md]: Can you put all the `do's` above the `don'ts`?

File: CODE_OF_CONDUCT.md (hunk 13 above)
Comment (jschipp-r7) [reply CODE_OF_CONDUCT.md]: We're using the code of conduct from an existing Rapid7 project.

Comment (mrinehart-r7) [reply CODE_OF_CONDUCT.md]: 👍 I get that - but can you confirm THEY had legal check off on it?

File: CONTRIBUTING.md (hunk 16 above)
Comment (jschipp-r7) [reply CONTRIBUTING.md]: Reordered

File: CONTRIBUTING.md (hunk 15 above)
Comment (jschipp-r7) [reply CONTRIBUTING.md]: Updated

File: CODE_OF_CONDUCT.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+By adopting this Code of Conduct, project maintainers commit themselves to
+fairly and consistently applying these principles to every aspect of managing
+this project. Project maintainers who do not follow or enforce the Code of
+Conduct may be permanently removed from the project team.
+
+This Code of Conduct applies both within project spaces and in public spaces
+when an individual is representing the project or its community.
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be
+reported by contacting the project maintainers at msfdev@metasploit.com. If
```
Comment (mrinehart-r7) [inline CODE_OF_CONDUCT.md]: Wrong email add
```suggestion
reported by contacting the project maintainers at IntegrationAlliance@rapid7.com. If
```

File: CODE_OF_CONDUCT.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+the incident involves a committer, you may report directly to
+jonathan_schipp@rapid7.com.
```
Comment (mrinehart-r7) [inline CODE_OF_CONDUCT.md]: I think this should still point to IntegrationAlliance@rapid7.com. Outside of that, I'm not sure you want to put your personal work email out like this.

Comment (jschipp-r7) [reply CODE_OF_CONDUCT.md]: My thoughts are that it should go to the manager, to not distract engineers.

Comment (mrinehart-r7) [reply CODE_OF_CONDUCT.md]: Fair point. Can you get an email alias setup for that purpose?

File: CODE_OF_CONDUCT.md (hunk 17 above)
Comment (jschipp-r7) [reply CODE_OF_CONDUCT.md]: Good catch here

File: CODE_OF_CONDUCT.md (hunk 13 above)
Comment (jschipp-r7) [reply CODE_OF_CONDUCT.md]: @aanand-r7 will be reviewing. Good point.

**Comment by amenon-r7** [comment]
The opening line... I'd change it to the following

Thank you for your interest in joining the InsightConnect Developer community!

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,78 @@
+# Contributing
+
+Thank you for your interest in improving InsightConnect! Please review our [Code of Conduct] before making contributions.
```
Comment (amenon-r7) [inline CONTRIBUTING.md]: Thank you for your interest in joining the InsightConnect Developer community!

Comment (jschipp-r7) [reply CONTRIBUTING.md]: Nice!

Comment (jschipp-r7) [reply CONTRIBUTING.md]: Updated

File: CODE_OF_CONDUCT.md (hunk 13 above)
Comment (jschipp-r7) [reply CODE_OF_CONDUCT.md]: Approval from @aanand-r7  after adding this: https://github.com/rapid7/insightconnect-plugins/pull/4/commits/cf4bf91aa4ab2bc6f6c33f2f09bd03764b82ae6b

Screen Shot 2019-06-10 at 10 51 30 AM

**Pull request merged**